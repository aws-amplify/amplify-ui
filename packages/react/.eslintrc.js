// TODO remove these once the full set of rules can be turned on for the repo
// extensions that should be ran against the entire repo
const sharedExtensions = [
  'plugin:react-hooks/recommended',

  // always extend last to override previous extensions
  'prettier',
];

// rules that should be ran against the entire repo
const sharedRules = {
  // react hook rules
  'react-hooks/rules-of-hooks': 'error',
  'react-hooks/exhaustive-deps': 'error',
};

module.exports = {
  env: { node: true },
  root: true,
  ignorePatterns: [
    // top level directories
    '__tests__',
    'dist',
    'node_modules',
    'scripts',

    // config files
    '.eslintrc.js',
    'jest.config.js',
    'jest.setup.ts',
    'rollup.config.js',
    'tsup.config.ts',

    // TODO remove once the full set of rules can be turned on for the repo
    'src/react-shim.js',
  ],
  extends: sharedExtensions,
  plugins: ['react-hooks'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: { jsx: true },
    ecmaVersion: 12,
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
    sourceType: 'module',
  },
  overrides: [
    {
      files: ['**/*.spec.*', '**/*.test.*'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        project: ['./tsconfig.dev.json'],
      },
      plugins: ['@typescript-eslint', 'jest'],
      rules: {
        // Turn the original rule off for test files, turn on the jest rule
        '@typescript-eslint/unbound-method': 'off',
        'jest/unbound-method': 'error',
      },
    },
    // TODO remove once the full set of rules can be turned on for the repo
    // To keep the rules-of-hooks on for the entire ui-react package while
    // we add the larger set of rules incrementally, use an overrides object
    // to incrementally apply the new rules
    {
      files: [
        // TODO Uncomment the below lines on an individual basis as the updates to pass linting are completed.
        // If needed, these can be broken down further, e.g. 'src/primitives/Alert/**/*'
        'src/components/**/*',
        'src/helpers/**/*',
        'src/hooks/**/*',
        'src/primitives/+(shared|utils)/**/*',
        'src/studio',
        // 'src/primitives/**/*',
      ],
      extends: [
        'eslint:recommended',
        'plugin:jest/recommended',
        'plugin:react/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',

        // extend last to override previous extensions
        ...sharedExtensions,
      ],
      plugins: ['@typescript-eslint', 'jest', 'react', 'react-hooks'],
      settings: {
        react: {
          version: 'detect',
        },
      },
      rules: {
        ...sharedRules,

        // typescript eslint rules either not in recommended rule set or overridden
        '@typescript-eslint/explicit-module-boundary-types': 2,
        '@typescript-eslint/member-ordering': 'error',
        '@typescript-eslint/no-extra-semi': 'error',
        '@typescript-eslint/no-floating-promises': ['off'],
        '@typescript-eslint/no-unused-expressions': [
          'error',
          { allowTernary: true },
        ],
        '@typescript-eslint/no-use-before-define': ['error'],
        '@typescript-eslint/no-unused-vars': [
          'error',
          { argsIgnorePattern: '_', varsIgnorePattern: '_' },
        ],
        '@typescript-eslint/prefer-nullish-coalescing': 'error',
        '@typescript-eslint/restrict-template-expressions': ['off'],
        '@typescript-eslint/unbound-method': 'error',

        // eslint rules either not in recommended rule set or overridden
        'comma-dangle': ['error', 'only-multiline'],
        'function-paren-newline': 'off',
        'generator-star-spacing': 'off',
        'global-require': 'off',
        'implicit-arrow-linebreak': 'off',
        'max-params': 2,
        'no-alert': 'error',
        'no-console': 'error',
        'no-eval': 'error',
        'no-tabs': ['error', { allowIndentationTabs: true }],
        'no-unused-vars': 'off', // prefer @typescript-eslint version
        'prefer-const': 'error',
        'prefer-destructuring': ['error', { array: false, object: true }],

        // jest rules either not in recommended rule set or overridden
        'jest/expect-expect': ['error', { assertFunctionNames: ['expect*'] }],
        'jest/no-mocks-import': 'off',

        // react rules either not in recommended rule set or overridden
        'react/destructuring-assignment': ['error', 'always'],
        'react/jsx-boolean-value': 'error',
        'react/jsx-no-constructed-context-values': ['error'],
        'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
        'react/jsx-props-no-spreading': 'off',
        'react/jsx-wrap-multilines': ['error', { declaration: 'ignore' }],
        'react/no-array-index-key': 'off',
        'react/no-danger': ['error'],
        'react/no-unused-prop-types': ['error'],
        'react/prop-types': 'off',
        'react/static-property-placement': ['error', 'static public field'],
      },
    },
  ],

  rules: sharedRules,
};
