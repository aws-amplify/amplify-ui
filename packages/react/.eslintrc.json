{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "extends": ["plugin:react-hooks/recommended"],
  "rules": {
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "error"
  },
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "ignorePatterns": ["scripts", "dist", "node_modules"]
}
