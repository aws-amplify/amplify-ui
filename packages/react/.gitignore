dist
# ignore generated primitive catalog ESM
src/PrimitiveCatalog.ts 

# ignore jest generated coverage directory
coverage 
