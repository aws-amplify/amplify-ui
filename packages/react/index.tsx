import { I18n } from '@aws-amplify/core';
import { dict } from '../core';

export * from './src';

I18n.putVocabularies(dict);
