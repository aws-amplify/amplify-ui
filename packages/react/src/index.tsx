export * from './components';
export * from './hooks';
export * from './primitives';

export * as components from './components';
export * as primitives from './primitives';

export { theme } from './defaultTheme';
