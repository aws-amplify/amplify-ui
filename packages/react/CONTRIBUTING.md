# Contributing

## Getting Started

1. Navigate to the _root_ of your local clone of [aws-amplify/amplify-ui](https://github.com/aws-amplify/amplify-ui)
1. Run `yarn setup`
1. Run `yarn react dev`

This will start building `@aws-amplify/ui-react` in watch mode. To test your changes, you can utilize [`examples/next`](../../examples/next) to run examples on `next.js`. Please see examples [README](../../examples/README.md) and e2e [README](../e2e/README.md#contributing) to get started.

## Dependencies

`@aws-amplify/ui-react` depends on [`@aws-amplify/ui`](../ui) for theming, state management, and translation logic. If you're looking for change in these, please refer to `@aws-amplify/ui` [README](../ui/README.md).

## Code Standards

### Imports

Separate all imports into organized, alphabetical blocks of `third_party → internal → local` for easier reading.

```js
import { isEmpty } from 'lodash/isEmpty';
import * as React from 'react';

import { Auth } from 'aws-amplify';

import { Button } from './primitives/Button';
import { THIS_ENUM } from './utils/types';
```

Do **NOT** use implicit paths like below. This can lead to circular dependencies unintentionally which is bad for tree shaking.

```js
import { Flex, Heading } from '../../..';
```

Use explicit import paths instead, the more specific the better.

```js
import { Flex } from '../../../primitives/Flex';
import { Heading } from '../../../primitives/Heading';
```

### Variable Naming

1. Component names should be capitalized.
2. Boolean props should be prefixed with `is` and default to `false`.

### Others

Do **NOT** introduce any [side effects](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) unless you have to with rational reasons. A "side effect" is defined as code that performs a special behavior when imported, other than exposing one or more exports. An example of this are polyfills, which affect the global scope and usually do not provide an export. For example:

```js
import './polyfill';
```

Any imported file is subject to tree shaking, that means CSS file needs to be added to the side effect list so it will not be unintentionally dropped:

```js
import './style.css';
```

## Testing for Production

After you tested your change, you can run `yarn react build` from monorepo to run build for production.
