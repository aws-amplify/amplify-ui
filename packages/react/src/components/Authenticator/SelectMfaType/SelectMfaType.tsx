import * as React from 'react';

import { authenticatorTextUtil } from '@aws-amplify/ui';
import { useAuthenticator } from '@aws-amplify/ui-react-core';

import { Flex } from '../../../primitives/Flex';
import { Heading } from '../../../primitives/Heading';
import { RadioGroupField } from '../../../primitives/RadioGroupField';
import { Radio } from '../../../primitives/Radio';
import { useCustomComponents } from '../hooks/useCustomComponents';
import { useFormHandlers } from '../hooks/useFormHandlers';
import { ConfirmSignInFooter } from '../shared/ConfirmSignInFooter';
import { RemoteErrorMessage } from '../shared/RemoteErrorMessage';
import { RouteContainer, RouteProps } from '../RouteContainer';

const {
  getMfaTypeLabelByValue,
  getSelectMfaTypeByChallengeName,
  getSelectMfaTypeText,
} = authenticatorTextUtil;

export const SelectMfaType = ({
  className,
  variation,
}: RouteProps): JSX.Element => {
  const { isPending, allowedMfaTypes = [] } = useAuthenticator((context) => {
    return [context.isPending, context.allowedMfaTypes];
  });

  const { handleChange, handleSubmit } = useFormHandlers();

  const {
    components: {
      // @ts-ignore
      SelectMfaType: {
        Header = SelectMfaType.Header,
        Footer = SelectMfaType.Footer,
      },
    },
  } = useCustomComponents();

  return (
    <RouteContainer className={className} variation={variation}>
      <form
        data-amplify-form=""
        data-amplify-authenticator-select-mfa-type=""
        method="post"
        onChange={handleChange}
        onSubmit={handleSubmit}
      >
        <Flex as="fieldset" direction="column" isDisabled={isPending}>
          <Header />

          <Flex direction="column">
            <RadioGroupField
              name="mfa_type"
              legend={getSelectMfaTypeText()}
              legendHidden
              isDisabled={isPending}
              isRequired
            >
              {allowedMfaTypes.map((value, index) => (
                <Radio
                  name="mfa_type"
                  key={value}
                  value={value}
                  defaultChecked={index === 0}
                >
                  {getMfaTypeLabelByValue(value)}
                </Radio>
              ))}
            </RadioGroupField>

            <RemoteErrorMessage />
          </Flex>

          <ConfirmSignInFooter />
          <Footer />
        </Flex>
      </form>
    </RouteContainer>
  );
};

SelectMfaType.Header = function Header(): JSX.Element {
  const { challengeName } = useAuthenticator((context) => {
    return [context.challengeName];
  });

  return (
    <Heading level={3}>
      {getSelectMfaTypeByChallengeName(challengeName)}
    </Heading>
  );
};

SelectMfaType.Footer = function Footer(): JSX.Element {
  // @ts-ignore
  return null;
};
