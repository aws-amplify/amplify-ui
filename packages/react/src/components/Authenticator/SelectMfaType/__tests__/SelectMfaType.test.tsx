import * as React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import { useAuthenticator, UseAuthenticator } from '@aws-amplify/ui-react-core';
import { AuthenticatorServiceFacade } from '@aws-amplify/ui';

import { SelectMfaType } from '../SelectMfaType';

jest.mock('@aws-amplify/ui-react-core');

jest.mock('../../hooks/useCustomComponents', () => ({
  useCustomComponents: () => ({
    components: {
      Header: () => null,
      Footer: () => null,
      SelectMfaType: { Header: () => null, Footer: () => null },
    },
  }),
}));

const fieldName = 'mfa_type';
const emailRadioLabel = 'Email Message';
const totpRadioLabel = 'Authenticator App (TOTP)';
const emailFieldInput = { name: fieldName, value: 'EMAIL' };
const totpFieldInput = { name: fieldName, value: 'TOTP' };

const mockUpdateForm = jest.fn();
const mockSubmitForm = jest.fn();
const mockToSignIn = jest.fn();

const mockUseAuthenticator = jest.mocked(useAuthenticator);

const mockUseAuthenticatorOutput: Partial<UseAuthenticator> = {
  authStatus: 'authenticated',
  challengeName: 'SELECT_MFA_TYPE',
  error: undefined as unknown as AuthenticatorServiceFacade['error'],
  route: 'selectMfaType',
  submitForm: mockSubmitForm,
  toSignIn: mockToSignIn,
  updateForm: mockUpdateForm,
  allowedMfaTypes: ['EMAIL', 'TOTP'],
  validationErrors: {} as AuthenticatorServiceFacade['validationErrors'],
};

mockUseAuthenticator.mockReturnValue(mockUseAuthenticatorOutput as any);

const props = {
  className: '',
  variation: 'default' as const,
};

describe('SelectMfaType', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected ', () => {
    const { container } = render(<SelectMfaType {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('sends change event on form input', async () => {
    render(<SelectMfaType {...props} />);

    const totpRadioButton = await screen.findByText(totpRadioLabel);

    fireEvent.click(totpRadioButton);

    expect(mockUpdateForm).toHaveBeenCalledWith(totpFieldInput);

    const emailRadioButton = await screen.findByText(emailRadioLabel);

    fireEvent.click(emailRadioButton);

    expect(mockUpdateForm).toHaveBeenCalledWith(emailFieldInput);
  });

  it('sends submit event on form submit with default', async () => {
    render(<SelectMfaType {...props} />);

    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });

    fireEvent.click(submitButton);

    expect(mockSubmitForm).toHaveBeenCalledTimes(1);
    expect(mockSubmitForm).toHaveBeenCalledWith({
      [emailFieldInput.name]: emailFieldInput.value,
    });
  });

  it('sends submit event on form submit with selection', async () => {
    render(<SelectMfaType {...props} />);

    const totpRadioButton = await screen.findByText(totpRadioLabel);

    fireEvent.click(totpRadioButton);

    expect(mockUseAuthenticatorOutput.updateForm).toHaveBeenCalledWith(
      totpFieldInput
    );

    const submitButton = await screen.findByRole('button', { name: 'Confirm' });

    fireEvent.click(submitButton);

    expect(mockSubmitForm).toHaveBeenCalledTimes(1);
    expect(mockSubmitForm).toHaveBeenCalledWith({
      [totpFieldInput.name]: totpFieldInput.value,
    });
  });

  it('displays error if present', async () => {
    mockUseAuthenticator.mockReturnValue({
      ...mockUseAuthenticatorOutput,
      error: 'mockError',
    } as any);

    render(<SelectMfaType {...props} />);

    expect(await screen.findByText('mockError')).toBeInTheDocument();
  });

  it('handles back to sign in button as expected', async () => {
    render(<SelectMfaType {...props} />);

    const backToSignInButton = await screen.findByRole('button', {
      name: 'Back to Sign In',
    });

    fireEvent.click(backToSignInButton);

    expect(mockToSignIn).toHaveBeenCalledTimes(1);
  });

  it('disables the submit button if confirmation is pending', async () => {
    mockUseAuthenticator.mockReturnValue({
      ...mockUseAuthenticatorOutput,
      isPending: true,
    } as any);

    render(<SelectMfaType {...props} />);

    const submitButton = await screen.findByRole('button', {
      name: 'Confirming',
    });

    expect(submitButton).toBeDisabled();
  });
});
