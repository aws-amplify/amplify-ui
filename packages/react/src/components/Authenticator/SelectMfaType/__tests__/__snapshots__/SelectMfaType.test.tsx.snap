// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMfaType renders as expected  1`] = `
<div>
  <div
    class=""
    data-amplify-authenticator=""
    data-variation="default"
  >
    <div
      data-amplify-container=""
    >
      <div
        data-amplify-router=""
      >
        <form
          data-amplify-authenticator-select-mfa-type=""
          data-amplify-form=""
          method="post"
        >
          <fieldset
            class="amplify-flex"
            style="flex-direction: column;"
          >
            <div
              class="amplify-flex"
              style="flex-direction: column;"
            >
              <fieldset
                class="amplify-flex amplify-fieldset amplify-fieldset--plain amplify-field amplify-radiogroupfield"
                role="radiogroup"
              >
                <legend
                  class="amplify-visually-hidden"
                >
                  Select MFA Type
                </legend>
                <div
                  aria-hidden="true"
                  class="amplify-fieldset__legend amplify-visually-hidden"
                >
                  Select MFA Type
                </div>
                <div
                  class="amplify-flex amplify-radiogroup"
                  id="amplify-id-:r0:"
                >
                  <label
                    class="amplify-flex amplify-radio"
                  >
                    <span
                      class="amplify-text amplify-radio__label"
                    >
                      Email Message
                    </span>
                    <input
                      aria-invalid="false"
                      checked=""
                      class="amplify-input amplify-field-group__control amplify-visually-hidden amplify-radio__input"
                      name="mfa_type"
                      required=""
                      type="radio"
                      value="EMAIL"
                    />
                    <span
                      aria-hidden="true"
                      class="amplify-flex amplify-radio__button"
                    />
                  </label>
                  <label
                    class="amplify-flex amplify-radio"
                  >
                    <span
                      class="amplify-text amplify-radio__label"
                    >
                      Authenticator App (TOTP)
                    </span>
                    <input
                      aria-invalid="false"
                      class="amplify-input amplify-field-group__control amplify-visually-hidden amplify-radio__input"
                      name="mfa_type"
                      required=""
                      type="radio"
                      value="TOTP"
                    />
                    <span
                      aria-hidden="true"
                      class="amplify-flex amplify-radio__button"
                    />
                  </label>
                </div>
              </fieldset>
            </div>
            <div
              class="amplify-flex"
              style="flex-direction: column;"
            >
              <button
                class="amplify-button amplify-field-group__control amplify-button--primary"
                type="submit"
              >
                Confirm
              </button>
              <button
                class="amplify-button amplify-field-group__control amplify-button--link amplify-button--small"
                type="button"
              >
                Back to Sign In
              </button>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  </div>
</div>
`;
